jax/numpy/vectorize.py

```diff
@@ -0,0 +1,236 @@
+            "output must be a tuple when multiple outputs are expected, "
+            "got: {!r}\n{}".format(out, error_context))
+      if len(out_shapes) != len(output_core_dims):
+        raise TypeError(
+            'wrong number of output arguments: expected %r, got %r %s'
+            % (len(output_core_dims), len(out_shapes), error_context))
+
+    sizes = dict(dim_sizes)
+    for shape, core_dims in zip(out_shapes, output_core_dims):
+      _update_dim_sizes(sizes, shape, core_dims, error_context,
+                        is_input=False)
+
+    return out
+  return wrapped
+
+
+def _apply_excluded(func, excluded, args):
+  if not excluded:
+    return func, args
+
+  if max(excluded) >= len(args):
+    raise ValueError("excluded={!r} is invalid for {!r} argument(s)"
+                     .format(excluded, len(args)))
+
+  dynamic_args = [arg for i, arg in enumerate(args) if i not in excluded]
+  static_args = [(i, args[i]) for i in sorted(excluded)]
+
+  def new_func(*args):
+    args = list(args)
+    for i, arg in static_args:
+      args.insert(i, arg)
+    return func(*args)
+
+  return new_func, dynamic_args
+
+
+@_wraps(onp.vectorize, lax_description=textwrap.dedent("""
+    JAX's implementation of vectorize should be considerably more efficient
+    than NumPy's, because it uses a batching transformation rather than an
+    explicit "for" loop.
+
+    Note that JAX only supports the optional ``excluded`` (integer only) and
+    ``signature`` arguments, both of which must be specified with keywords.
+    """))
+def vectorize(pyfunc, *, excluded=frozenset(), signature=None):
+
+  if any(not isinstance(exclude, int) for exclude in excluded):
+    raise TypeError("jax.numpy.vectorize can only exclude integer arguments, "
+                    "but excluded={!r}".format(excluded))
+  if excluded and min(excluded) < 0:
+    raise ValueError("excluded={!r} contains negative numbers".format(excluded))
+
+  @functools.wraps(pyfunc)
+  def vectorized(*args):
+    error_context = ("on vectorized function with excluded={!r} and "
+                     "signature={!r}".format(excluded, signature))
+    excluded_func, args = _apply_excluded(pyfunc, excluded, args)
+    args = tuple(map(np.asarray, args))
+
+    if signature is not None:
+      input_core_dims, output_core_dims = _parse_gufunc_signature(signature)
+      broadcast_args, dim_sizes = _broadcast_with_core_dims(
+          args, input_core_dims, error_context)
+    else:
+      input_core_dims = [()] * len(args)
+      broadcast_args = np.broadcast_arrays(*args)
+      output_core_dims = None
+      dim_sizes = {}
+
+    checked_func = _check_output_dims(
+        excluded_func, dim_sizes, output_core_dims, error_context)
+
+    num_batch_dims = broadcast_args[0].ndim - len(input_core_dims[0])
+    vectorized_func = checked_func
+    for _ in range(num_batch_dims):
+      vectorized_func = api.vmap(vectorized_func)
+    return vectorized_func(*broadcast_args)
```

Comment:
It occurs to me that there is an alternative implementation here that could potentially be more efficient: instead of broadcasting everything and vmapping everything, we could only broadcast if needed and apply vmap with select batched arguments instead. This could potentially be significantly more efficient in case where partial batching rules exist, e.g., as was implemented in https://github.com/google/jax/pull/2138.

Reply:
Done.